[SPARK-14369][SQL] Locality support for FileScanRDD

liancheng opened the pull request:
## What changes were proposed in this pull request?

This PR adds preliminary locality support for `FileFormat` data sources by overriding `FileScanRDD.preferredLocations()`. The strategy can be divided into two parts:
1.  Block location lookup
   
   Unlike `HadoopRDD` or `NewHadoopRDD`, `FileScanRDD` doesn't have access to the underlying `InputFormat` or `InputSplit`, and thus can't rely on `InputSplit.getLocations()` to gather locality information. Instead, this PR queries block locations using `FileSystem.getBlockLocations()` after listing all `FileStatus`es in `HDFSFileCatalog` and convert all `FileStatus`es into `LocatedFileStatus`es.
   
   Note that although S3/S3A/S3N file systems don't provide valid locality information, their `getLocatedStatus()` implementations don't actually issue remote calls either. So there's no need to special case these file systems.
2.  Selecting preferred locations
   
   For each `FilePartition`, we pick up top 3 locations that containing the most data to be retrieved. This isn't necessarily the best algorithm out there. Further improvements may be brought up in follow-up PRs.
## How was this patch tested?

Tested by overriding default `FileSystem` implementation for `file:///` with a mocked one, which returns mocked block locations.

liancheng committed:
Locality support for FileScanRDD

SparkQA commented:
**[Test build #54859 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54859/consoleFull)** for PR 12153 at commit [`179ff58`](https://github.com/apache/spark/commit/179ff5898cf1aa2acc52514e1703b0534b5b6d7c).

SparkQA commented:
**[Test build #54859 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54859/consoleFull)** for PR 12153 at commit [`179ff58`](https://github.com/apache/spark/commit/179ff5898cf1aa2acc52514e1703b0534b5b6d7c).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54859/
Test FAILed.

liancheng committed:
Fixes import order

SparkQA commented:
**[Test build #54866 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54866/consoleFull)** for PR 12153 at commit [`9a68989`](https://github.com/apache/spark/commit/9a689890b32d1c11c401a5aa7b0ff77c0b377c21).

JoshRosen commented:
FYI, I don't think you can necessarily put multiple `test-hadoop` commands in the same PR title and have it run multiple tests.

liancheng commented:
@JoshRosen Oh, thanks for reminding. Then I'll test different Hadoop versions manually by changing PR titles.

SparkQA commented:
**[Test build #54866 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54866/consoleFull)** for PR 12153 at commit [`9a68989`](https://github.com/apache/spark/commit/9a689890b32d1c11c401a5aa7b0ff77c0b377c21).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54866/
Test FAILed.

liancheng committed:
Fixes cases where block location list is empty

SparkQA commented:
**[Test build #54873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54873/consoleFull)** for PR 12153 at commit [`e846fd1`](https://github.com/apache/spark/commit/e846fd132f057e283a51cfcc3632f936474b46f1).

SparkQA commented:
**[Test build #54873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54873/consoleFull)** for PR 12153 at commit [`e846fd1`](https://github.com/apache/spark/commit/e846fd132f057e283a51cfcc3632f936474b46f1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54873/
Test PASSed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #54910 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54910/consoleFull)** for PR 12153 at commit [`e846fd1`](https://github.com/apache/spark/commit/e846fd132f057e283a51cfcc3632f936474b46f1).

SparkQA commented:
**[Test build #54910 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54910/consoleFull)** for PR 12153 at commit [`e846fd1`](https://github.com/apache/spark/commit/e846fd132f057e283a51cfcc3632f936474b46f1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54910/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #54981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54981/consoleFull)** for PR 12153 at commit [`e846fd1`](https://github.com/apache/spark/commit/e846fd132f057e283a51cfcc3632f936474b46f1).

liancheng committed:
Adds withHadoopConf utility method

SparkQA commented:
**[Test build #54981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54981/consoleFull)** for PR 12153 at commit [`e846fd1`](https://github.com/apache/spark/commit/e846fd132f057e283a51cfcc3632f936474b46f1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54981/
Test PASSed.

liancheng committed:
Tests

SparkQA commented:
**[Test build #54988 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54988/consoleFull)** for PR 12153 at commit [`a1f527a`](https://github.com/apache/spark/commit/a1f527aa2c91776085818c360f9ea3d0a8d5d616).

SparkQA commented:
**[Test build #54988 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54988/consoleFull)** for PR 12153 at commit [`a1f527a`](https://github.com/apache/spark/commit/a1f527aa2c91776085818c360f9ea3d0a8d5d616).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54988/
Test PASSed.

liancheng committed:
Moves the test case to HadoopFsRelationTest

SparkQA commented:
**[Test build #54998 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54998/consoleFull)** for PR 12153 at commit [`5fef611`](https://github.com/apache/spark/commit/5fef61170f5b63845f8b5ee72fb3413e1ce4477d).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala:
```diff
@@ -620,17 +621,28 @@ class HDFSFileCatalog(
         }
+
+        fs match {
+          case _: S3FileSystem =>
```
S3AFileSystem / NativeS3FileSystem. S3FileSystem is for the block store.

yhuai replied:
Actually, do we need to have this pattern matching? Looks like S3 filesystems just use the default implementation of getFileBlockLocations, which returns `localhost:50010`?

liancheng replied:
Thanks, made corresponding changes and added comments.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala:
```diff
@@ -620,17 +621,28 @@ class HDFSFileCatalog(
+
+          case _ =>
+            statuses.map { f =>
+              new LocatedFileStatus(f, fs.getFileBlockLocations(f, 0, f.getLen))
```
At here, let's check if this file status is already a LocatedFileStatus.

liancheng replied:
Done.

SparkQA commented:
**[Test build #54998 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54998/consoleFull)** for PR 12153 at commit [`5fef611`](https://github.com/apache/spark/commit/5fef61170f5b63845f8b5ee72fb3413e1ce4477d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LocalityTestFileSystem extends RawLocalFileSystem `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54998/
Test FAILed.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala:
```diff
@@ -86,4 +87,8 @@ class FileScanRDD(
   override protected def getPartitions: Array[Partition] = filePartitions.toArray
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    split.asInstanceOf[FilePartition].files.flatMap(_.locations).distinct
```
Looks we want to pick the location based on the size of a block? In MapOutputTracker we have `getPreferredLocationsForShuffle`, which can be related to us.

liancheng replied:
Added a constraint so that we only pick locations of blocks whose sizes are larger than 10% of the whole file partition.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:
```diff
@@ -137,10 +140,12 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
                 val size = if (remaining > maxSplitBytes) maxSplitBytes else remaining
-                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size)
+                val hosts = getBlockHosts(blockLocations, offset)
+                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size, hosts)
```
How about we add document to this FileSourceStrategy to explain our strategy? So, readers can quickly understand what we are doing at here.

liancheng replied:
Added.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:
```diff
@@ -205,4 +210,20 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
+    case f => Array.empty[BlockLocation]
+  }
+
+  private def getBlockIndex(blockLocations: Array[BlockLocation], offset: Long): Int = {
```
What is the semantic of `offset`?

liancheng replied:
Added comments.

liancheng committed:
Addresses PR comments

SparkQA commented:
**[Test build #55118 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55118/consoleFull)** for PR 12153 at commit [`4004501`](https://github.com/apache/spark/commit/40045015bbfb9f4e0699534271565efb458b73c9).

liancheng committed:
Adjusts file partition size fraction threshold to 10%

SparkQA commented:
**[Test build #55120 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55120/consoleFull)** for PR 12153 at commit [`0c3882f`](https://github.com/apache/spark/commit/0c3882fddb00aa2edb01e0db6f7b16cfb7952a33).

SparkQA commented:
**[Test build #55118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55118/consoleFull)** for PR 12153 at commit [`4004501`](https://github.com/apache/spark/commit/40045015bbfb9f4e0699534271565efb458b73c9).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55118/
Test FAILed.

SparkQA commented:
**[Test build #55120 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55120/consoleFull)** for PR 12153 at commit [`0c3882f`](https://github.com/apache/spark/commit/0c3882fddb00aa2edb01e0db6f7b16cfb7952a33).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55120/
Test FAILed.

SparkQA commented:
**[Test build #55225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55225/consoleFull)** for PR 12153 at commit [`17aad3e`](https://github.com/apache/spark/commit/17aad3e7debdf6c163ab5cdd76df25794a41bbca).

SparkQA commented:
**[Test build #55225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55225/consoleFull)** for PR 12153 at commit [`17aad3e`](https://github.com/apache/spark/commit/17aad3e7debdf6c163ab5cdd76df25794a41bbca).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55225/
Test FAILed.

liancheng committed:
Fixes test case flakiness

SparkQA commented:
**[Test build #55228 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55228/consoleFull)** for PR 12153 at commit [`3a0e2f7`](https://github.com/apache/spark/commit/3a0e2f71e9ea147083d4d26167fc58ce0b3323cb).

nongli commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:
```diff
@@ -137,10 +140,14 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
+                // Finds out a list of location hosts where lives the first block that contains the
+                // starting point the file block starts at `offset` with length `size`.
+                val hosts = getBlockHosts(blockLocations, offset)
+                PartitionedFile(partition.values, file.getPath.toUri.toString, offset, size, hosts)
```
How is this locality information maintained when we coalesce in the logic below?

liancheng replied:
Each `FilePartition` derives their own preferred location in `FileScanRDD.preferredLocations()` from location hosts of all its `PartitionedFile`s.

nongli replied:
This is not likely to generate good locality right?

Is it reasonable to repurpose this?
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala#L124

liancheng replied:
This is a really good pointer, thanks!

davies replied:
We could do that in a separate PR (improve the locality with bounded complicity).

SparkQA commented:
**[Test build #55228 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55228/consoleFull)** for PR 12153 at commit [`3a0e2f7`](https://github.com/apache/spark/commit/3a0e2f71e9ea147083d4d26167fc58ce0b3323cb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55228/
Test FAILed.

liancheng committed:
Uses getLocatedStatus() instead of getStatus() + getBlockLocations()

SparkQA commented:
**[Test build #55529 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55529/consoleFull)** for PR 12153 at commit [`739420b`](https://github.com/apache/spark/commit/739420b20b4cd398b4b43f3f5a3db36f4023cdb3).

SparkQA commented:
**[Test build #55529 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55529/consoleFull)** for PR 12153 at commit [`739420b`](https://github.com/apache/spark/commit/739420b20b4cd398b4b43f3f5a3db36f4023cdb3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55529/
Test FAILed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala:
```diff
@@ -86,4 +87,20 @@ class FileScanRDD(
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    val files = split.asInstanceOf[FilePartition].files
+    val partitionSize = files.map(_.length).sum
```
how about:
```
val hostToLength = Map[String, Long]()
for (file <- files) {
  for (host <- file.hosts) {
    if (hostToLength.contains(host)) {
      hostToLength(host) += file.length
    } else {
      hostToLength(host) = file.length
    }
  }
  hostToLength.toSeq.sortBy(_._2).take(3).map(_._1)
}
```

liancheng committed:
Fixes import order

liancheng committed:
Makes locality test case more robust

liancheng committed:
Simplifies locality selection

SparkQA commented:
**[Test build #55713 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55713/consoleFull)** for PR 12153 at commit [`cef40bf`](https://github.com/apache/spark/commit/cef40bfb940a24e43ac01e4c521be6de51f5662d).

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala:
```diff
@@ -668,6 +671,42 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
+          .option("dataSchema", df1.schema.json)
+          .load(path)
+
+        val Some(fileScanRDD) = {
```
how about:
```
val fileScanRDD = df2.queryExecution.executedPlan.collectFirst {
  case scan: DataSourceScan if scan.rdd.isInstanceOf[FileScanRDD] => scan.rdd.asInstanceOf[FileScanRDD]
}
```

liancheng replied:
Yea, this is better, thanks!

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala:
```diff
@@ -668,6 +671,42 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
+          }
+        }
+
+        val partitions = fileScanRDD.partitions
```
do we always return 2 partitions?

liancheng replied:
Not necessarily, and probably only 1, since both files written are pretty small and will be merged into a single partition.

yhuai replied:
If we expect them be packed in the same partition, let's add an `assert`.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:
```diff
@@ -197,4 +204,26 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
+  // Given locations of all blocks of a single file, `blockLocations`, and an `offset` position of
+  // the same file, find out the index of the block that contains this `offset`. If no such block
+  // can be found, returns -1.
+  private def getBlockIndex(blockLocations: Array[BlockLocation], offset: Long): Int = {
```
This method is only called once, should we inline it?

liancheng replied:
This one is a little bit tricky, would like to keep it here for readability.

cloud-fan commented:
Overall LGTM

davies commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala:
```diff
@@ -86,4 +87,9 @@ class FileScanRDD(
+
+  override protected def getPreferredLocations(split: Partition): Seq[String] = {
+    val files = split.asInstanceOf[FilePartition].files
+    if (files.isEmpty) Seq.empty else files.maxBy(_.length).locations
```
Should we pick the top 3 host that hold most of the data?

liancheng replied:
Updated. @cloud-fan also suggested the same strategy. Thanks!

SparkQA commented:
**[Test build #55713 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55713/consoleFull)** for PR 12153 at commit [`cef40bf`](https://github.com/apache/spark/commit/cef40bfb940a24e43ac01e4c521be6de51f5662d).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55713/
Test FAILed.

davies commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:
```diff
@@ -197,4 +204,26 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
+  // Given locations of all blocks of a single file, `blockLocations`, and an `offset` position of
+  // the same file, find out the block that contains this `offset`, and returns location hosts of
+  // that block. If no such block can be found, returns an empty array.
+  private def getBlockHosts(blockLocations: Array[BlockLocation], offset: Long): Array[String] = {
```
Should we pick the hosts based on range that the host hold most of data in that range?

liancheng replied:
Makes sense, updated.

liancheng committed:
Better locality selection strategy

- When picking block locations for a file segment, pick the block
  that containing the largest portion of the given file segment.

- When picking preferred locations for a given FilePartition, pick top
  3 hosts that containing the most data to be retrieved.

SparkQA commented:
**[Test build #55812 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55812/consoleFull)** for PR 12153 at commit [`c372a90`](https://github.com/apache/spark/commit/c372a90f2fc03e116f095c7cd1dce534b47104cc).

SparkQA commented:
**[Test build #55813 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55813/consoleFull)** for PR 12153 at commit [`63613ef`](https://github.com/apache/spark/commit/63613ef2d42ec93f0b466116c702316f06fbb7d6).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala:
```diff
@@ -624,20 +624,40 @@ class HDFSFileCatalog(
 
   def getStatus(path: Path): Array[FileStatus] = leafDirToChildrenFiles(path)
 
+  private implicit class LocatedFileStatusIterator(iterator: RemoteIterator[LocatedFileStatus])
```
Why do we need this?

liancheng replied:
It's used [here][1]. `RemoteIterator` doesn't extend from `Iterator`. This implicit conversion helps to convert contents behind a `RemoteIterator` into an array.

[1]: https://github.com/apache/spark/pull/12153/files#diff-40c347747af9101e7e9fee52fc4120b8R656

yhuai commented on sql/hive/src/test/scala/org/apache/spark/sql/sources/hadoopFsRelationSuites.scala:
```diff
@@ -668,6 +671,42 @@ abstract class HadoopFsRelationTest extends QueryTest with SQLTestUtils with Tes
     }
   }
+
+  test("Locality support for FileScanRDD") {
```
How about we add the following test cases?

* Every partition of the FileScanRDD has a single file.
* There are some large files. So, there are some partitions and every such a partition has a part of a file.
* If we are going to pick the top K locations that host the largest amount of data, let's also add tests.

SparkQA commented:
**[Test build #55812 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55812/consoleFull)** for PR 12153 at commit [`c372a90`](https://github.com/apache/spark/commit/c372a90f2fc03e116f095c7cd1dce534b47104cc).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55812/
Test FAILed.

SparkQA commented:
**[Test build #55813 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55813/consoleFull)** for PR 12153 at commit [`63613ef`](https://github.com/apache/spark/commit/63613ef2d42ec93f0b466116c702316f06fbb7d6).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55813/
Test FAILed.

liancheng commented:
retest this please

SparkQA commented:
**[Test build #55834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55834/consoleFull)** for PR 12153 at commit [`63613ef`](https://github.com/apache/spark/commit/63613ef2d42ec93f0b466116c702316f06fbb7d6).

SparkQA commented:
**[Test build #55834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55834/consoleFull)** for PR 12153 at commit [`63613ef`](https://github.com/apache/spark/commit/63613ef2d42ec93f0b466116c702316f06fbb7d6).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55834/
Test FAILed.

yhuai commented:
@liancheng Add more tests?

liancheng committed:
More tests

liancheng commented:
retest this please

liancheng commented:
Seems that Jenkins is down...

SparkQA commented:
**[Test build #56090 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56090/consoleFull)** for PR 12153 at commit [`3484559`](https://github.com/apache/spark/commit/34845596b508715762f76374005f7e0e89db4ff0).

SparkQA commented:
**[Test build #56090 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56090/consoleFull)** for PR 12153 at commit [`3484559`](https://github.com/apache/spark/commit/34845596b508715762f76374005f7e0e89db4ff0).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56090/
Test FAILed.

yhuai commented:
test this please

SparkQA commented:
**[Test build #56158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56158/consoleFull)** for PR 12153 at commit [`3484559`](https://github.com/apache/spark/commit/34845596b508715762f76374005f7e0e89db4ff0).

SparkQA commented:
**[Test build #56158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56158/consoleFull)** for PR 12153 at commit [`3484559`](https://github.com/apache/spark/commit/34845596b508715762f76374005f7e0e89db4ff0).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56158/
Test FAILed.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala:
```diff
@@ -621,20 +621,40 @@ class HDFSFileCatalog(
+
   private def listLeafFiles(paths: Seq[Path]): mutable.LinkedHashSet[FileStatus] = {
     if (paths.length >= sqlContext.conf.parallelPartitionDiscoveryThreshold) {
       HadoopFsRelation.listLeafFilesInParallel(paths, hadoopConf, sqlContext.sparkContext)
```
Seems we also need to update the `listLeafFiles` that is called by `listLeafFilesInParallel`.

yhuai replied:
Let's also have a test.

liancheng committed:
Collects block locations during parallel file listing

SparkQA commented:
**[Test build #56234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56234/consoleFull)** for PR 12153 at commit [`86f1195`](https://github.com/apache/spark/commit/86f1195c3f114aaf3a01f9043302379c09f0d755).

yhuai commented:
Seems it is hanging in BucketedReadSuite again... Let's look at it together and figure out what's wrong.

SparkQA commented:
**[Test build #56234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56234/consoleFull)** for PR 12153 at commit [`86f1195`](https://github.com/apache/spark/commit/86f1195c3f114aaf3a01f9043302379c09f0d755).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class FakeBlockLocation(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56234/
Test FAILed.

liancheng committed:
Workarounds the deadlock issue

SparkQA commented:
**[Test build #56354 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56354/consoleFull)** for PR 12153 at commit [`79beca6`](https://github.com/apache/spark/commit/79beca6ad09df76a39cc1a80eebed378e3b2fffc).

liancheng commented:
All the timeout in the Jenkins builds were due to a deadlock in `DAGScheduler`, and can be steadily reproduced locally by running the following test case

> BucketedReadSuite.only shuffle one side when 2 bucketed tables have different bucket keys.

This test case creates two bucketed tables both with 8 buckets and then joins them. Reducing 8 to 5 eliminates the deadlock. But I haven't figured out the real reason behind the deadlock. The deadlock also disappears if I remove FileScanRDD.preferredLocations(). Maybe that too many tasks are scheduled to the same place and exhausted some thread-pool?

liancheng commented:
I'm closing this one for #12527, which is a rebased version of this one.

SparkQA commented:
**[Test build #56354 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56354/consoleFull)** for PR 12153 at commit [`79beca6`](https://github.com/apache/spark/commit/79beca6ad09df76a39cc1a80eebed378e3b2fffc).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56354/
Test PASSed.
